src/AI/JakeMan.java

```diff
@@ -0,0 +1,770 @@
+    }
+
+    if( shouldResupply )
+    {
+      log(String.format("  %s needs supplies.", unit.toStringWithLocation()));
+      goals.addAll(stations);
+      if( avoidProduction )
+        goals.removeAll(AIUtils.findAlliedIndustries(gameMap, myArmy, goals, !avoidProduction));
+    }
+    else if( uc.possibleActions.contains(UnitActionFactory.CAPTURE) )
+    {
+      for( XYCoord xyc : futureCapTargets )
+        if( !AIUtils.isCapturing(gameMap, myArmy.cos[0], xyc) )
+          goals.add(xyc);
+    }
+    else if( uc.possibleActions.contains(UnitActionFactory.ATTACK) )
+    {
+      Map<UnitModel, Double> valueMap = new HashMap<UnitModel, Double>();
+      Map<UnitModel, ArrayList<XYCoord>> targetMap = new HashMap<UnitModel, ArrayList<XYCoord>>();
+
+      // Categorize all enemies by type, and all types by how well we match up vs them
+      final ArrayList<XYCoord> allEnemies = AIUtils.findEnemyUnits(myArmy, gameMap);
+      Utils.sortLocationsByTravelTime(unit, allEnemies, gameMap);
+      for( XYCoord xyc : allEnemies )
+      {
+        Unit target = gameMap.getResident(xyc);
+        UnitModel model = target.model;
+        XYCoord targetCoord = new XYCoord(target.x, target.y);
+        if (Utils.findShortestPath(unit, targetCoord, gameMap, true) != null &&
+            isWeakTo(target.model, unit.model))
+        {
+          valueMap.put(model, (double)target.getCost());
+          if (!targetMap.containsKey(model)) targetMap.put(model, new ArrayList<XYCoord>());
+          targetMap.get(model).add(targetCoord);
+        }
+      }
+
+      // Sort all individual target lists by distance
+      for (ArrayList<XYCoord> targetList : targetMap.values())
+        Utils.sortLocationsByTravelTime(unit, targetList, gameMap);
+
+      // Sort all target types by how much we want to shoot them with this unit
+      Queue<Entry<UnitModel, Double>> targetTypesInOrder = 
+          new PriorityQueue<Entry<UnitModel, Double>>(myArmy.cos[0].unitModels.size(), new UnitModelFundsComparator());
+      targetTypesInOrder.addAll(valueMap.entrySet());
+
+      while (!targetTypesInOrder.isEmpty())
+      {
+        UnitModel model = targetTypesInOrder.poll().getKey(); // peel off the juiciest
+        goals.addAll(targetMap.get(model)); // produce a list ordered by juiciness first, then distance TODO: consider a holistic "juiciness" metric that takes into account both matchup and distance?
+      }
+    }
+
+    if( goals.isEmpty() ) // Send 'em at production facilities if they haven't got anything better to do
+    {
+      for( XYCoord coord : futureCapTargets )
+      {
+        MapLocation loc = gameMap.getLocation(coord);
+        if( unit.CO.unitProductionByTerrain.containsKey(loc.getEnvironment().terrainType)
+            && myArmy.isEnemy(loc.getOwner()) )
+        {
+          goals.add(coord);
+        }
+      }
+    }
+
+    if( goals.isEmpty() ) // If there's really nothing to do, go to MY HQ
+      goals.addAll(myArmy.HQLocations);
```

Comment:
Why go to its own HQ instead of the enemy HQ locations?

Reply:
Because those should have been contemplated already.

Reply:
Ah, touché. So this is definitely a "we should never actually get here" kind of scenario. I normally like to throw in a warning when that happens, if for no other reason than that it's always fun when I see it pop up.